app/code/Magento/Catalog/Model/Indexer/Category/Flat/AbstractAction.php

```diff
@@ -77,20 +76,31 @@ class AbstractAction
      */
     protected $skipStaticColumns = [];
 
+    /**
+     * @var SkipStaticColumnsProvider
+     */
+    private $staticColumnsProvider;
+
     /**
      * @param ResourceConnection $resource
      * @param StoreManagerInterface $storeManager
      * @param Helper $resourceHelper
+     * @param MetadataPool|null $categoryMetadata
+     * @param SkipStaticColumnsProvider|null $provider
      */
     public function __construct(
         ResourceConnection $resource,
         StoreManagerInterface $storeManager,
-        Helper $resourceHelper
+        Helper $resourceHelper,
+        MetadataPool $categoryMetadata = null,
+        SkipStaticColumnsProvider $provider = null
     ) {
         $this->resource = $resource;
         $this->connection = $resource->getConnection();
         $this->storeManager = $storeManager;
         $this->resourceHelper = $resourceHelper;
+        $this->categoryMetadata = $categoryMetadata ?: ObjectManager::getInstance()->get(MetadataPool::class);
+        $this->staticColumnsProvider = $provider ?: ObjectManager::getInstance()->get(SkipStaticColumnsProvider::class);
```

Comment:
Please use null-coalescing operator in constructor.
Also, some of the logic have been lost, which is why we have failed Integration Tests

```suggestion
        $metadataPool = $metadataPool ?? ObjectManager::getInstance()->get(MetadataPool::class);
        $this->categoryMetadata = $metadataPool->getMetadata(CategoryInterface::class);
        $this->skipStaticColumnsProvider = $skipStaticColumnsProvider
            ?? ObjectManager::getInstance()->get(SkipStaticColumnsProvider::class);
```